### ISSUE TITLE
zef is throwing an error when trying to install Module::Name instead of Module-Name

### ISSUE THREAD

**Issue by Altai-man** [ugexe/zef#139]
On latest zef(0.1.3). There are `Perl6::Parser` module, the name added in META.json is `Perl6::Parser`. Although, when you try to install it with `zef install Perl6::Parser` zef goes like that:

```
➜  ~ zef install Perl6::Parser;
===> Searching for: Perl6::Parser
===> Testing: Perl6-Parser
t/01-parse-text.t .......... ok 
t/02-validate.t ............ ok 
t/03-build-tree.t .......... ok 
t/04-parse.t ............... ok 
t/10-multi-declarator.t .... ok 
t/11-package-declarator.t .. ok   
t/12-regex-declarator.t .... ok 
t/13-scope-declarator.t .... ok 
t/14-type-declarator.t ..... ok 
t/15-routine-declarator.t .. ok 
t/16-quote.t ............... ok   
t/999-regression.t ......... ok  
t/comments.t ............... ok 
t/declaration.t ............ ok 
t/operators.t .............. ok   
t/pair.t ................... ok  
t/pod.t .................... ok 
t/regex.t .................. ok 
t/rosetta-1.t .............. ok 
t/rosetta-a.t .............. ok   
t/rosetta-b.t .............. ok   
t/single-term.t ............ ok 
t/subroutine.t ............. ok 
All tests successful.
Files=23, Tests=230,  409 wallclock secs
Result: PASS
===> Testing [OK] for Perl6-Parser
===> Testing: Perl6::Parser
t/01-parse-text.t .......... ok 
t/02-validate.t ............ ok 
t/03-build-tree.t .......... ok 
t/04-parse.t ............... ok 
t/05-iterator.t ............ ok  
t/06-editing.t ............. ok 
t/10-multi-declarator.t .... ok 
t/11-package-declarator.t .. ok   
t/12-regex-declarator.t .... ok 
t/13-scope-declarator.t .... ok 
t/14-type-declarator.t ..... ok 
t/15-routine-declarator.t .. ok 
t/16-quote.t ............... ok  
t/999-regression.t ......... ok  
t/comments.t ............... ok 
t/corpus/rosetta-1.t ....... ok 
t/corpus/rosetta-a.t ....... ok   
t/corpus/rosetta-b.t ....... ok   
t/declaration.t ............ ok 
t/operators.t .............. ok   
t/pair.t ................... ok  
t/pod.t .................... ok 
t/regex.t .................. ok 
t/single-term.t ............ ok 
t/subroutine.t ............. ok 
All tests successful.
Files=25, Tests=261,  215 wallclock secs
Result: PASS
===> Testing [OK] for Perl6::Parser
===> Installing: Perl6-Parser
===> Installing: Perl6::Parser
===> Install [FAIL] for Perl6::Parser: ===SORRY!=== Error while compiling /home/sena/site#sources/5D35B70B6DCF6175A88D947FEB7B35E06D3EA5EB (Perl6::Parser)
Undeclared routine:
    and used at lines 399, 406. Did you mean 'rand', 'any', 'end'?


===SORRY!=== Error while compiling /home/sena/site#sources/5D35B70B6DCF6175A88D947FEB7B35E06D3EA5EB (Perl6::Parser)
Undeclared routine:
    and used at lines 399, 406. Did you mean 'rand', 'any', 'end'?


  in block  at /home/sena/.rakudobrew/moar-nom/install/share/perl6/site/sources/B70C99508AF99FBB1C38AC562646FC3CE8ED23DC (Zef::Client) line 556
  in block  at /home/sena/.rakudobrew/moar-nom/install/share/perl6/site/sources/B70C99508AF99FBB1C38AC562646FC3CE8ED23DC (Zef::Client) line 519
  in sub  at /home/sena/.rakudobrew/moar-nom/install/share/perl6/site/sources/B70C99508AF99FBB1C38AC562646FC3CE8ED23DC (Zef::Client) line 508
  in method install at /home/sena/.rakudobrew/moar-nom/install/share/perl6/site/sources/B70C99508AF99FBB1C38AC562646FC3CE8ED23DC (Zef::Client) line 600
  in sub MAIN at /home/sena/.rakudobrew/moar-nom/install/share/perl6/site/sources/0C5228A6A32D31F472FE88164A1943264F5A699F (Zef::CLI) line 123
  in block <unit> at /home/sena/.rakudobrew/moar-nom/install/share/perl6/site/resources/A7381E13F425B98E260D5430380344243C6A39E3 line 1
```
Obviously, it does testing twice for `Perl6-Parser` and for `Perl6::Parser`(but on the second run there are more tests for some reason), and there is a weird error on attempt to install `Perl6::Parser` - undeclared `and` routine.

But if you type `zef install Perl6-Parser`, it will be installed and tested totally fine:
```
➜  ~ zef install Perl6-Parser
===> Searching for: Perl6-Parser
===> Testing: Perl6-Parser
t/01-parse-text.t .......... ok 
t/02-validate.t ............ ok 
t/03-build-tree.t .......... ok 
t/04-parse.t ............... ok 
t/10-multi-declarator.t .... ok 
t/11-package-declarator.t .. ok  
t/12-regex-declarator.t .... ok 
t/13-scope-declarator.t .... ok 
t/14-type-declarator.t ..... ok 
t/15-routine-declarator.t .. ok 
t/16-quote.t ............... ok   
t/999-regression.t ......... ok  
t/comments.t ............... ok 
t/declaration.t ............ ok 
t/operators.t .............. ok   
t/pair.t ................... ok   
t/pod.t .................... ok 
t/regex.t .................. ok 
t/rosetta-1.t .............. ok 
t/rosetta-a.t .............. ok   
t/rosetta-b.t .............. ok   
t/single-term.t ............ ok 
t/subroutine.t ............. ok 
All tests successful.
Files=23, Tests=230,  439 wallclock secs
Result: PASS
===> Testing [OK] for Perl6-Parser
===> Installing: Perl6-Parser
➜  ~ 
```

Latest rakudo distrubiton:
```
➜ ~ perl6 -v
This is Rakudo version 2017.02-247-g8574322 built on MoarVM version 2017.02-38-g5845784
```

I'm not sure is it zef or this module issue, though, but I guess it should be investigated.

**Comment by ugexe** [comment]
Hmm I suspect this is due to:

1) The name changing `s/::/-/` (nothing wrong with this by itself, but...)
2) Perl6::Parser uses the version `*` before and after this name change.
3) zef has a local cache, and if the versions from all sources are the same (and they technically are: `*`), it will choose the cache to install from (but why its choosing BOTH is a mystery)
4) The install error is from rakudo itself during install (this does not happen in zef), and is happening during precompile. Its possible that rakudo is mixing files from these two different versions (because the META6 *says* they are the same).

---

Having never installed Perl6::Parser myself previously (so its not in my zef cache) the zef install went as expected. So I would think if you run `zef nuke RootDir` (which is the same as `rm -rf ~/.zef`) and run that command again it will work.

*but* before you do this, would it be possible for you to zip/tar your `~/.zef` directory (with `--exclude=**/.git/*`) and send it to me, as well as show me the output of `RAKUDO_MODULE_DEBUG=1 ZEF_PLUGIN_DEBUG=1 zef --debug install Perl6::Parser`? This will help me get to the bottom of why zef is attempting to install two distributions (instead of either the outdated one in cache, or the newer one just fetched from the ecosystem - both the same but with a different name `s/-/::/`).

---

To summarize:

* The zef bug here is that it tries to install two distribution candidates to fulfill a single request
* The rakudo bug here is that precompilation seems to be confused by having multiple `Perl6::Parser:ver<*>` installed (here I am referring to the value in META6 provides field, the module itself not the distribution name) that have different source (since the distributions have different names rakudo doesn't die when trying to install both).

**Comment by Altai-man** [comment]
>show me the output of RAKUDO_MODULE_DEBUG=1 ZEF_PLUGIN_DEBUG=1 zef --debug install Perl6::Parser

Do you want me to do it after I do `zef uninstall Perl6-Parser` and `zef uninstall Perl6::Parser` or when it's already installed?

**Comment by ugexe** [comment]
Yeah after, or `zef nuke site` (deletes all installed modules, including zef, but might be necessary to uninstall due to this naming issue). I essentially want to see what your original output would have been with those two commands.

**Comment by Altai-man** [comment]
[zef.zip](https://github.com/ugexe/zef/files/845167/zef.zip)
[debug-output.txt](https://github.com/ugexe/zef/files/845191/debug-output.txt) by `RAKUDO_MODULE_DEBUG=1 ZEF_PLUGIN_DEBUG=1 zef --debug install Perl6::Parser &> debug-output`.
Hope that helps, ask for anything that can be helpful from my side.

**Comment by ugexe** [comment]
Thanks for those references. The `zef` bug of installing two distributions for a single request where they both have different distribution names but a matching module name in provides is fixed.

Note that Perl6::Parser won't install currently because the author has a typo in their source-url, but if they fix that it will install as expected.
